### PR TITLE
Add `Resource` and `ResourceManager` classes 

### PR DESCRIPTION
**Pull request by jakemac53**
Fixes https://github.com/dart-lang/build/issues/396

See [README.md](https://github.com/dart-lang/build/pull/410/files#diff-ecbc6c47898b2960c58173697b2fb550R163) for more info.

### CODE REVIEW

File: build/README.md [hunk 1]
```diff
@@ -160,6 +160,42 @@ class ResolvingCopyBuilder {
...
+The `Resource<T>` constructor takes a single required argument which is a
+factory function that returns a `FutureOr<T>`. There is also a named argument
+`dispose` which is called at the end of life for the resource, with the
+instance that should be disposed. This returns a `FutureOr<dynamic>`.
+
+So a simple example `Resource` would look like this:
+
+```dart
+final resource = new Resource(
+  () => createMyExpensiveResource(),
```
Comment (natebosch) [inline build/README.md]: [nit] unnecessary closure?

File: build/lib/src/resource/resource.dart [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+/// Build system implementations should be the only users that directly
+/// instantiate a [ResourceManager] since they can handle the lifecycle
+/// guarantees in a sane way.
+class Resource<T> {
+  /// Factory method which creates an instance of this resource.
+  final FutureOr<T> Function() _create;
+
+  /// Optional method which is given an existing instance that is ready to be
+  /// disposed.
+  final FutureOr Function(T instance) _userDispose;
```
Comment (natebosch) [inline build/lib/src/resource/resource.dart]: Stable SDK analyzer still chokes on this. Need a private typedef for the field and the old function syntax in any method signatures.

File: build/lib/src/resource/resource.dart [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    return _instance;
+  }
+
+  /// Disposes the actual instance of this resource if present.
+  Future _dispose() {
+    if (_instance == null) return new Future.value(null);
+    var oldInstance = _instance;
+    _instance = null;
+    if (_userDispose != null) {
+      return oldInstance.then((old) => _userDispose(old));
```
Comment (natebosch) [inline build/lib/src/resource/resource.dart]: [nit] unnecessary closure

File: build/test/generate/run_builder_test.dart [hunk 4]
```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
```
Comment (natebosch) [inline build/test/generate/run_builder_test.dart]: increment the year for new files

File: build/README.md (hunk 1 above)
Comment (jakemac53) [reply build/README.md]: for the docs though I think this is more clear what the argument is (unless I actually showed a fake impl of createExpensiveResource which I think is not necessary)

Comment (natebosch) [reply build/README.md]: fair enough

File: build/lib/src/resource/resource.dart (hunk 2 above)
Comment (jakemac53) [reply build/lib/src/resource/resource.dart]: done

File: build/lib/src/resource/resource.dart (hunk 3 above)
Comment (jakemac53) [reply build/lib/src/resource/resource.dart]: done

File: build/test/generate/run_builder_test.dart (hunk 4 above)
Comment (jakemac53) [reply build/test/generate/run_builder_test.dart]: done

**Pull request merged**